### PR TITLE
DR2-1331 Create E2E tests

### PR DESCRIPTION
**Pull request by MancunianSam**
This includes:
* 4 Tests for ingesting judgments
* A GitHub action to deploy to ECR
* An action to run the tests
* An action to run scalafmt
A Dockerfile to build the image.

### CODE REVIEW

File: .github/scripts/run_ecs_task.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+
+
+def run_tests():
+    subnets_response = ec2_client.describe_subnets(
+        Filters=[{'Name': 'tag:Name', 'Values': ['intg-vpc-private-subnet-eu-west-2a']}]
+    )
+    security_groups_response = ec2_client.describe_security_groups(
+        Filters=[{'Name': 'tag:Name', 'Values': ['intg-outbound-https']}]
+    )
+    group_ids = [x["GroupId"] for x in security_groups_response["SecurityGroups"]]
```
Comment (techncl) [inline .github/scripts/run_ecs_task.py]: Could `x` be renamed to `securityGroup` or `secGroup`?

File: .github/scripts/run_ecs_task.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    subnet_ids = [x["SubnetId"] for x in subnets_response["Subnets"]]
```
Comment (techncl) [inline .github/scripts/run_ecs_task.py]: Could `x` be renamed to `subnet`?

File: .github/scripts/run_ecs_task.py [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+        launchType='FARGATE',
+        networkConfiguration={
+            'awsvpcConfiguration': {
+                'subnets': subnet_ids,
+                'securityGroups': group_ids
+            }
+        },
+        taskDefinition='e2e-tests'
+    )
+    task_arns = [x["taskArn"] for x in response["tasks"]]
```
Comment (techncl) [inline .github/scripts/run_ecs_task.py]: Could `x` be renamed to `task`?

File: .github/workflows/run.yml [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+          aws-region: eu-west-2
+          role-session-name: RunTests
+      - run: |
+          pip install boto3
+          python ./.github/scripts/run_ecs_task.py
+      - name: Send failure message
+        uses: nationalarchives/dr2-github-actions/.github/actions/slack-send@main
+        if: failure()
+        with:
+          message: ":warning: E2E tests have failed"
```
Comment (techncl) [inline .github/workflows/run.yml]: Is it possible to link to the PR or if not, to the "Actions" tab?

File: project/build.properties [hunk 5]
```diff
@@ -0,0 +1 @@
+sbt.version=1.9.6
```
Comment (techncl) [inline project/build.properties]: There is 1.9.7 now

File: src/test/scala/uk/gov/nationalarchives/JudgmentSpec.scala [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+import org.scalatest.concurrent.Eventually
+import org.scalatest.flatspec.AnyFlatSpec
+import org.scalatest.matchers.should.Matchers._
+import org.scalatest.time.{Minutes, Span}
+
+import java.util.UUID
+
+class JudgmentSpec extends AnyFlatSpec with GivenWhenThen with ParallelTestExecution with Eventually {
+  def urlWithoutCourt(): String = {
+    val id = (Math.random * 1000000).toInt.toString
```
Comment (techncl) [inline src/test/scala/uk/gov/nationalarchives/JudgmentSpec.scala]: You don't need to convert this to a String since you are using string interpolation below

File: src/test/scala/uk/gov/nationalarchives/JudgmentSpec.scala [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+    s"https://example.com/id/2023/$id"
+  }
+  def validUri(): String = {
+    val id = (Math.random * 1000000).toInt.toString
```
Comment (techncl) [inline src/test/scala/uk/gov/nationalarchives/JudgmentSpec.scala]: You don't need to convert this to a String since you are using string interpolation below

File: project/Dependencies.scala [hunk 8]
```diff
@@ -0,0 +1,16 @@
+import sbt._
+
+object Dependencies {
+  lazy val docx4j = "org.docx4j" % "docx4j" % "6.1.2"
+  lazy val jaxb = "org.glassfish.jaxb" % "jaxb-runtime" % "2.3.7"
+  lazy val circeCore = "io.circe" %% "circe-core" % "0.14.5"
+  lazy val circeGeneric = "io.circe" %% "circe-generic" % "0.14.5"
```
Comment (techncl) [inline project/Dependencies.scala]: Could the version be in a variable for these two?

File: project/Dependencies.scala [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+  lazy val s3Client = "uk.gov.nationalarchives" %% "da-s3-client" % "0.1.27"
+  lazy val sqsClient = "uk.gov.nationalarchives" %% "da-sqs-client" % "0.1.27"
```
Comment (techncl) [inline project/Dependencies.scala]: Could the version be in a variable for these two?

File: project/Dependencies.scala [hunk 10]
```diff
@@ -0,0 +1,16 @@
...
+  lazy val preservicaClient = "uk.gov.nationalarchives" %% "preservica-client-fs2" % "0.0.28"
+  lazy val log4jCore = "org.apache.logging.log4j" % "log4j-core" % "2.20.0"
+  lazy val log4jSlf4j = "org.apache.logging.log4j" % "log4j-slf4j-impl" % "2.20.0"
+  lazy val log4jJson = "org.apache.logging.log4j" % "log4j-layout-template-json" % "2.20.0"
```
Comment (techncl) [inline project/Dependencies.scala]: Could the version be in a variable for these three?

File: src/test/scala/uk/gov/nationalarchives/Steps.scala [hunk 11]
```diff
@@ -0,0 +1,182 @@
...
+import uk.gov.nationalarchives.dp.client.fs2.Fs2Client.{contentClient, entityClient}
+
+import java.security.MessageDigest
+import java.time.OffsetDateTime
+import java.util.UUID
+
+object Steps {
+  private val url = sys.env("PRESERVICA_API_URL")
+  private val secretName = sys.env("PRESERVICA_SECRET_NAME")
+  private val environment = sys.env.getOrElse("ENVIRONMENT", "intg")
```
Comment (techncl) [inline src/test/scala/uk/gov/nationalarchives/Steps.scala]: Do we need a `.getOrElse` here? If there's no environment, wouldn't it be better to throw an error? Also, I was thinking if you're using the sandbox env, and you forget to add an env, maybe the error message might be confusing?

File: src/test/scala/uk/gov/nationalarchives/JudgmentSpec.scala [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+    And("The SourceId identifiers will be set correctly")
+    val identifiers = getIdentifiers(entity.parent)
+    identifiers.size should equal(1)
+    identifiers("SourceID") should equal(uri)
+  }
+
+  "A judgment with a valid uri and a valid cite" should "ingest into Preservica" in {
+    Given("A judgment with a valid URI and a valid cite")
+    val uri = validUri()
+    val reference = createJudgment(Option("ewca"), Option(uri))
```
Comment (techncl) [inline src/test/scala/uk/gov/nationalarchives/JudgmentSpec.scala]: 1. It's not that big of a deal but for the cite, you have the court rather than an example of a cite `[{year}] {Court} {number} ({potential sub court})`
2. You use the term "valid cite" but shouldn't it just be `"A judgment with a valid URI and a cite"`?

File: src/test/scala/uk/gov/nationalarchives/JudgmentSpec.scala [hunk 13]
```diff
@@ -0,0 +1,117 @@
...
+    identifiers("SourceID") should equal(uri)
+    identifiers("URI") should equal(uri)
+    identifiers("Cite") should equal("ewca")
+    identifiers("Code") should equal("ewca")
+  }
+
+  "A judgment with a invalid court in the uri and the skipSeriesLookup parameter set" should "ingest into Preservica in the Unknown folder" in {
+    Given("A judgment with a valid URI but an invalid court and an invalid cite")
+    val uri = validUri().replace("ewca", "invalid")
+    val reference = createJudgment(Option("invalid"), Option(uri))
```
Comment (techncl) [inline src/test/scala/uk/gov/nationalarchives/JudgmentSpec.scala:77]: You use the term "invalid cite" but:

1. In the court handler code, we are not validating the cites, only the courts which we retrieve from the url (via Regex)
2. the cite in this test and the previous test are technically invalid cites since they are not in the format mentioned in my comment above

File: src/test/scala/uk/gov/nationalarchives/JudgmentSpec.scala [hunk 14]
```diff
@@ -0,0 +1,117 @@
...
+    identifiers.size should equal(3)
+    identifiers("SourceID") should equal("Court Documents (court not matched)")
+    identifiers("Cite") should equal("invalid")
+    identifiers("Code") should equal("invalid")
+  }
+
+  "A judgment with a missing court in the uri" should "ingest into Preservica in the Unknown folder" in {
+    Given("A judgment with a valid URI but an invalid court and an invalid cite")
+    val uri = urlWithoutCourt()
+    val reference = createJudgment(None, Option(uri))
```
Comment (techncl) [inline src/test/scala/uk/gov/nationalarchives/JudgmentSpec.scala:100]: Would this be a "null cite", rather than "invalid cite", like in the first test?

File: .github/workflows/test.yml [hunk 15]
```diff
@@ -0,0 +1,17 @@
+name: DR2 Run Lambda Tests
```
Comment (techncl) [inline .github/workflows/test.yml]: Are these still "Lambda" tests?

File: src/test/scala/uk/gov/nationalarchives/Steps.scala [hunk 16]
```diff
@@ -0,0 +1,182 @@
...
+    val tarBytes = tarFile(reference, docxBytes, metadataBytes)
+    for {
+      _ <- Logger[IO].info(s"Creating judgment with reference $reference")
+      reference <- Stream.emits[IO, Byte](tarBytes).chunks.map(_.toByteBuffer).toUnicastPublisher.use { publisher =>
+        daS3Client.upload(bucket, s"$reference.tar.gz", tarBytes.length, publisher).map(_ => reference)
+      }
+    } yield reference
+  }.unsafeRunSync()
+
+  def tarFile(reference: UUID, docxBytes: Array[Byte], metadataBytes: Array[Byte]): Array[Byte] = {
```
Comment (techncl) [inline src/test/scala/uk/gov/nationalarchives/Steps.scala]: Maybe `createTarFiles` or `tarFiles`?

File: README.md [hunk 17]
```diff
@@ -1 +1,28 @@
 # DR2 End to end tests
+
+These are tests of the DR2 ingest process. They only include tests cases that will be successfully ingested into
+Preservica.
+Test cases which will fail mid-process will be dealt with in unit tests on the relevant lambdas.
```
Comment (techncl) [inline README.md]: Missing comma after "mid-process"

File: README.md [hunk 18]
```diff
@@ -1 +1,28 @@
...
+
+## Structure
+
+The tests use the `GivenWhenThen` trait from ScalaTest. This allows us to write BDD style tests but allows us to run
```
Comment (techncl) [inline README.md]: Missing hyphen in between "BDD style"

File: README.md [hunk 19]
```diff
@@ -1 +1,28 @@
...
+
+These tests are packaged into a docker image which is run in ECS. 
+This allows us to access Preservica from inside the VPC. The task is triggered by a GitHub action defined
+in [run.yml](./.github/workflows/run.yml)
+
+## Environment variables
+
+| Variable Name          | Description                                       |
+|------------------------|---------------------------------------------------|
+| ACCOUNT_ID             | Used to construct the SQS queue URL               |
```
Comment (techncl) [inline README.md]: Maybe "AWS account Id for environment, used to construct the SQS queue URL"?

File: .gitignore [hunk 20]
```diff
@@ -0,0 +1,2 @@
+/.bsp/
```
Comment (techncl) [inline .gitignore]: Could you add some more things to ignore [like ](https://github.com/nationalarchives/dr2-ingest-parsed-court-document-event-handler/blob/main/.gitignore)?

**Review by techncl — COMMENTED** [review]
Looks good, just some minor comments, and...

I'm not sure if it's a silly question but do you think it's worth testing for:

1. duplicate judgments - we upload another judgment where everything's the same
2. duplicate judgments -  we upload another judgment where the name is the same
3. updated judgments

**Comment by MancunianSam** [comment]
I'm happy to add other tests in but I'm not sure how these would work:
1. duplicate judgments - we upload another judgment where everything's the same. What would we test for? That nothing has changed. 
2. duplicate judgments - we upload another judgment where the name is the same - Same here, what would we be checking for?
3. updated judgments - What is it we would update to change things. I suppose we could do the same url and different cite which would give a different id_Code?

The problem I think with these is that we would have to put two judgments in, a first one and then a second one once the first is done so you can check they've worked properly. lt's taking around 9 minutes to complete a single judgment. They're running in parallel so it's ok for now but if we have to put through two sequentially, it's going to be pushing 20 minutes for a run. 

I might have misunderstood but what do you think?

File: .github/scripts/run_ecs_task.py (hunk 1 above)
Comment (MancunianSam) [reply .github/scripts/run_ecs_task.py]: I've gone with `sec_group` as it's a bit more Python-y.

File: .github/workflows/run.yml (hunk 4 above)
Comment (MancunianSam) [reply .github/workflows/run.yml]: We can't do the PR because it's triggeredby workflow dispatch but we can do the workflow run. I've added it in.

File: src/test/scala/uk/gov/nationalarchives/Steps.scala (hunk 11 above)
Comment (MancunianSam) [reply src/test/scala/uk/gov/nationalarchives/Steps.scala]: It's a fair point. It was easier when I was testing but you're right. I've taken it out.

File: src/test/scala/uk/gov/nationalarchives/JudgmentSpec.scala (hunk 12 above)
Comment (MancunianSam) [reply src/test/scala/uk/gov/nationalarchives/JudgmentSpec.scala]: Makes sense, I've updated the cite and removed the word valid from the test name and the Given block.

File: src/test/scala/uk/gov/nationalarchives/JudgmentSpec.scala (hunk 13 above)
Comment (MancunianSam) [reply src/test/scala/uk/gov/nationalarchives/JudgmentSpec.scala:77]: I've changed it to `an invalid court and a non-standard format cite` That way, we're testing that whatever we put in the cite field gets put in the Cite and Code identifiers.

File: src/test/scala/uk/gov/nationalarchives/JudgmentSpec.scala (hunk 14 above)
Comment (MancunianSam) [reply src/test/scala/uk/gov/nationalarchives/JudgmentSpec.scala:100]: Yeah it is, I've changed it.

File: .github/workflows/test.yml (hunk 15 above)
Comment (MancunianSam) [reply .github/workflows/test.yml]: Nope, copy and paste fun. I've changed it to DR2 Run Scalafmt as that's all we're doing.

File: src/test/scala/uk/gov/nationalarchives/Steps.scala (hunk 16 above)
Comment (MancunianSam) [reply src/test/scala/uk/gov/nationalarchives/Steps.scala]: I've gone with `createTarFile` as we're only creating one at a time.

File: README.md (hunk 17 above)
Comment (MancunianSam) [reply README.md]: I think it's also right without it but I've put one in.

**Comment by techncl** [comment]
> I'm happy to add other tests in but I'm not sure how these would work:
> 
> 1. duplicate judgments - we upload another judgment where everything's the same. What would we test for? That nothing has changed.
> 2. duplicate judgments - we upload another judgment where the name is the same - Same here, what would we be checking for?
> 3. updated judgments - What is it we would update to change things. I suppose we could do the same url and different cite which would give a different id_Code?
> 
> The problem I think with these is that we would have to put two judgments in, a first one and then a second one once the first is done so you can check they've worked properly. lt's taking around 9 minutes to complete a single judgment. They're running in parallel so it's ok for now but if we have to put through two sequentially, it's going to be pushing 20 minutes for a run.
> 
> I might have misunderstood but what do you think?

- I think when I wrote 1 & 2, I was thinking that attempting to add a duplicate judgment to Preservica would throw an error/ignore the new attempt? Does it? I was thinking that if this is the case, then the test could ensure that this happened. 
     - To reduce the time taken, I was thinking, we could just have a judgment already in Preservica. 
     - Side note: Do we/should we have some sort of notification if this happens?
- > 3. updated judgments - What is it we would update to change things. I suppose we could do the same url and different cite which would give a different id_Code?
     - Yes, these; is it possible that we'd be asked to amend the title, also?
     - I was thinking, like the solution above, we could have a judgment already in Preservica

**Comment by MancunianSam** [comment]
I don't know if it throws an error or ignores the attempt. There's nothing in place to check for ingest errors yet anyway. I  know we might be able to check the process monitor but there's no guarantee that works properly so I'd rather only have successful judgments in for now. 

We could have a judgment already in Preservica but that means having a static named judgment in there which is liable to get deleted at any point. It has the potential to make the tests flaky.

I think considering the individual lambda unit tests are covering changes to things like Code, the fact that determining when an ingest has failed is tricky and that we're not 100% sure what we want to happen in some of these cases, I think it's fine to leave it as is. 

This was supposed to be a few skeleton happy path tests to get us going and they can be added to as and when.

**Comment by kurtismash** [comment]
Handing of duplicates is an interesting one that we haven't yet discussed or designed for but will need to do before we go live. As @techncl points out there's at least 2 different paths we need to handle; a duplicate event causing us to ingest the exact same thing, and a repeated action causing us to see the same document again, possibly with different metadata.

One of the reasons I've avoided this so far is that we have very little in the message incoming to us to identify a record or package of records. I'd like to handle this generically if possible, putting the handling within our Step Function rather than the handler. The ideal solution would be for each asset incoming to us to have an externally authored reference which we could lookup in Preservica before starting ingest. If it already exists we'd confirm the fixity of the original files match and tell everyone we've ingested it - [Idempotency](https://docs.aws.amazon.com/AWSEC2/latest/APIReference/Run_Instance_Idempotency.html#:~:text=Idempotency%20ensures%20that,current%20creation%20status.) at it's finest. If it doesn't exist, we ingest it. If there's a discrepancy, we alert.

Court Documents are an odd case where a reparse could produce new metadata that we've not seen before, we're not going to handle that though as it's really a cataloguing issue - we're making sure the documents are preserved and stop at that. If the same document was transferred to us multiple times, we'd preserve each instance of this document.

TL/DR; let's not worry about duplicates for now, it's something we need to do but requires a bit more thinking around it.

**Review by techncl — APPROVED** [review]
👍
